### DISCUSSION TITLE
Welcome to groupme Discussions!

### DISCUSSION THREAD

**Discussion by densestvoid** [densestvoid/groupme#9] [General]
## 👋 Welcome!
  We’re using Discussions as a place to connect with other members of our community. We hope that you:
  * Ask questions you’re wondering about.
  * Share ideas.
  * Engage with other community members.
  * Welcome others and are open-minded. Remember that this is a community we
  build together 💪.

  To get started, comment below with an introduction of yourself and tell us about what you do with this community.

**Comment by tekkamanendless** [comment]
Hello!  My primary use case right now is for integration with GroupMe for CAD Alert, an app that helps firefighters receive dispatches over GroupMe, among other things (web push, text messages, e-mail, etc.).  I'm currently rewriting it in Go (it's currently Java in production), and this package helped me keep things pretty simple.

My only work with GroupMe is limited to the example in #8: I create a temporary group, add a user to it by searching for her e-mail or phone number, extract the real user ID from the response, and then delete the group.  From there, I can DM the user since I know her user ID now.

In terms of this package, there are two things that I'll probably be submitting PRs for in the future:

1. The `HTTP_*` constants are redundant; the `http` package already provides constants for all of the status codes (for example, `http.StatusBadRequest`).  Since you're already using the `http` package, I recommend just using their constants.
2. `AddMembersResults` will basically never work on the first try; you pretty much always have to retry it until you stop getting a 503 error.  However, the way that you've structured the response doesn't really capture this.  The error that gets returned is just a plain-text error `fmt.Errorf("%s: %s", getResp.Status, string(bytes))`, which makes it difficult to process (and subject to random changes in how you decide to format the error).  Right now, I'm just checking that the string starts with `503 Service Unavailable`, but that's not very clean.  I'd rather do something like `if errors.Is(err, groupme.ErrServiceUnavailable)`, using some fixed error types (`ErrServiceUnavailable`, `ErrBadRequest`, etc.) and maybe wrapping them to include the body (similar to what you're doing now), such as `return fmt.Errorf("%w: %s", ErrServiceUnavailable, string(bytes))`.

**Reply by densestvoid** [reply]
That's awesome! Glad to hear the package is being used, and with a great purpose!

In terms of the future PRs:
1. I see that now. I am very open to and would happily accept the PR, but I would still like it to include the specific string values for each status, as the reasoning is sometimes specific to GroupMe and helpful for debugging. I'd also like to keep status code 420 as it is not in the http package but is a very useful return error. If you do decide to write a PR for this, feel free to open an issue and let me know how you'd approach this.

2. This does appear rather burdensome to check for errors, especially when looking for a specific code. While this is distinct from the above PR, they would probably go hand in hand. Instead of errors for specific codes, what if a generic groupme request error is returned with and http.Status and body? Let me know what you think.

Let me tell you, I really appreciate you using this package, and would love to start moving towards a stable 1.0 release. With more people like you letting it see some use and submitting fixes in PRs, that will happen sooner rather than later.

**Reply by tekkamanendless** [reply]
#10 standardizes the error workflow a bit without introducing any new signatures (and it fixes some implicit failures in that a `*Meta` would never be returned).

As for simplifying the client workflow, there are two major directions to take:

**Direction 1:**
The package provides helper functions to test errors.
This is generally really simple and makes it so that the user doesn't have to try type-casting our return values.

```
result, err := client.XXXXXXXX(...)
if groupme.IsServiceUnavailable(err) {
   // Handle this error.
} else if groupme.IsEnhanceYourCalm(err) {
   // Handle this error.
} else if err != nil {
   // Oh no, we can't handle this error.
}
```

Implementation idea:

```
func IsServiceUnavailable(err error) bool {
   if err == nil {
      return false
   }
   meta, okay := err.(*Meta)
   if !okay {
      return false
   }
   return meta.Code == http.StatusServiceUnavailable
}
```

**Direction 2:**
The package provides base error types and the user relies on `errors.Is` to check them.
This is also fairly simple, but it does mean that we have to do some work when we generate the errors to make sure that we wrap the appropriate base type, which seems a bit more work than it's worth.

```
result, err := client.XXXXXXXX(...)
if errors.Is(err, groupme.ErrServiceUnavailable) {
   // Handle this error.
} else if errors.Is(err, groupme.ErrEnhanceYourCalm) {
   // Handle this error.
} else if err != nil {
   // Oh no, we can't handle this error.
}
```

Implementation idea:

```
var ErrServiceUnavailable = errors.New("503 Service Unavailable")

[...] // in client.do
if getResp.StatusCode/100 > 2 {
   switch getResp.StatusCode {
   case http.StatusServiceUnavailable:
      return fmt.Errorf("%w: %v", ErrServiceUnavailable, meta.Errors)
   [...]
}
```

Direction 1 seems a lot simpler to me, but :shrug:

**Reply by densestvoid** [reply]
I'm inclined to lean towards option 2, for the sole reason that Go seems to be heading in this direction. However, this can be tabled until after the recent error StatusCode handling pull request is resolved.
